File: cpp/linked-list/hard/23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode dummy;
        auto cmp = [](ListNode* l, ListNode* r){return l->val > r->val;};
        std::priority_queue<ListNode*, std::vector<ListNode*>, decltype(cmp)> q(cmp);
        // 每个链表的头节点加入优先队列
        for (auto e : lists) {
            if (e != nullptr) {
                q.push(e);
            }
        }

        auto p = &dummy;
        while (!q.empty()) {
            p->next = q.top();
            p = p->next;
            q.pop();
            if (p->next) {
                q.push(p->next);
            }
        }
        return dummy.next;
    }
    /*
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }

        if (lists.size() < 2) {
            return lists[0];
        }
        
        ListNode* result = nullptr;
        for (auto e : lists) {
            result = merge2Lists(e, result);
        }

        return result;
    }

    ListNode* merge2Lists(ListNode* list1, ListNode* list2) {
        ListNode dummy;
        auto p = &dummy;
        auto p1 = list1;
        auto p2 = list2;
        while (p1 && p2) {
            if (p1->val < p2->val) {
                p->next = p1;
                p1 = p1->next;
            } else {
                p->next = p2;
                p2 = p2->next;
            }
            p = p->next;
        }

        if (p1) {
            p->next = p1;
        }

        if (p2) {
            p->next = p2;
        }

        return dummy.next;
    }
    */
};
// @lc code=end

File: golang/linked-list/hard/23.合并 K 个升序链表.go
/*
 * @lc app=leetcode.cn id=23 lang=golang
 * @lcpr version=30202
 *
 * [23] 合并 K 个升序链表
 */

package leetcode_solutions

import (
	"container/heap"
	"testing"
)

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type MinHeap []*ListNode

func (h MinHeap) Len() int {
	return len(h)
}

func (h MinHeap) Less(i, j int) bool {
	return h[i].Val < h[j].Val
}

func (h MinHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(*ListNode))
}

func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}

	dummy := &ListNode{}
	p := dummy

	h := &MinHeap{}
	heap.Init(h)

	for _, head := range lists {
		if head != nil {
			heap.Push(h, head)
		}
	}

	for h.Len() > 0 {
		node := heap.Pop(h).(*ListNode)
		p.Next = node
		if node.Next != nil {
			heap.Push(h, node.Next)
		}
		p = p.Next
	}

	return dummy.Next
}

// @lc code=end

func TestMergeKSortedLists(t *testing.T) {
	// your test code here

}

/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

*/
